### PR TITLE
Validate label names

### PR DESCRIPTION
**Pull request by fabxc**
@beorn7

### CODE REVIEW

**Comment by fabxc** [comment]
Will throw some tests in there later.

File: expfmt/decode.go [hunk 1]
```diff
@@ -107,7 +107,23 @@ type protoDecoder struct {
...
+		if m == nil {
+			continue
+		}
+		for _, l := range m.GetLabel() {
+			if l == nil {
+				continue
+			}
+			if !model.LabelNameRE.MatchString(l.GetName()) {
+				return fmt.Errorf("invalid label name %q", l.GetName())
+			}
```
Comment (beorn7) [inline expfmt/decode.go]: I'm 80% sure that protobuf can transport invalid UTF-8 strings (while JSON can't). So (just yesterday) I ran into a library function to check if a string is a valid UTF-8 string (on which we depend in our hashing function).

So I'd add
`if !utf8.ValidString(l.GetValue()) { ... } `

**Comment by beorn7** [comment]
Otherwise :+1:

File: expfmt/decode.go (hunk 1 above)
Comment (fabxc) [inline expfmt/decode.go]: That would have to be applied to metric names and label values too then, right?

Comment (beorn7) [inline expfmt/decode.go]: The regexp match will also catch invalid strings (as not-matching, that is). I tried that out.

**Comment by fabxc** [comment]
Added unicode validation for proto strings.

File: expfmt/decode.go [hunk 2]
```diff
@@ -107,7 +108,30 @@ type protoDecoder struct {
 // Decode implements the Decoder interface.
 func (d *protoDecoder) Decode(v *dto.MetricFamily) error {
 	_, err := pbutil.ReadDelimited(d.r, v)
-	return err
+	if err != nil {
+		return err
+	}
+	if !utf8.ValidString(v.GetName()) {
```
Comment (beorn7) [inline expfmt/decode.go]: We have a regexp somewhere for metric names. We should check against the regexp here (which will include checking for valid UTF-8).

Comment (fabxc) [inline expfmt/decode.go]: Not in common/model at least. Until now.

**Comment by fabxc** [comment]
Performance A/B test:

Fresh servers, 32 cores, 64GB RAM, ~2800 targets.

Without checks:
 ~524,000 samples/sec
load avg ~12.5

With checks:
 ~514,000 samples/sec
load avg ~18.8

Certainly a significant CPU impact. Slight reduction in samples/sec.
We will see if that changes over time. Probably the samples/sec are not sustainable anyway.

**Comment by fabxc** [comment]
Updated. A/B test running.

File: model/metric.go [hunk 3]
```diff
@@ -79,3 +83,19 @@ func (m Metric) Fingerprint() Fingerprint {
...
+// IsValidMetricName returns true iff name matches the pattern of MetricNameRE.
+func IsValidMetricName(n LabelValue) bool {
+	if len(n) == 0 {
+		return false
+	}
+	if !((n[0] >= 'a' && n[0] <= 'z') || (n[0] >= 'A' && n[0] <= 'Z') || n[0] == '_' || n[0] == ':') {
+		return false
+	}
+	for _, b := range n[1:] {
+		if !((b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || b == ':' || (b >= '0' && b <= '9')) {
```
Comment (beorn7) [inline model/metric.go]: I think the previous `if` block can be easily avoided by taking the index in the for loop and loop over the whole string (`for i, b := range n {`) and then expand the condition to `... || (b >= '0' && b <= '9' && i>0))`.

Same above.

**Comment by fabxc** [comment]
This version has no notable impact on CPU load or samples/sec. Seems good to go.

File: model/metric.go (hunk 3 above)
Comment (fabxc) [inline model/metric.go]: Less code but more checks per subsequent iterations. Probably doesn't matter either way.

Comment (beorn7) [inline model/metric.go]: The "more checks" only kick in for numbers, so pretty rarely. With the amount of replicated code (and potential of unintentional deviations between L92 and L96), I'd vote pretty clearly for the one line version.

**Pull request merged**